src/bsoncxx/builder/list.hpp

```diff
@@ -118,6 +133,54 @@ class list {
         }
     }
 };
+
+///
+/// A JSON-like builder for creating documents.
+///
+class document : public list {
+    using initializer_list_t = std::initializer_list<list>;
+
+   public:
+    ///
+    /// Creates an empty document.
+    ///
+    document() : list({}, false, false){};
+
+    ///
+    /// Creates a BSON document.
+    ///
+    /// @param init
+    ///     the initializer list used to construct the BSON document
+    ///
+    /// @see bsoncxx::builder::list
+    /// @see bsoncxx::builder::array
+    ///
+    document(initializer_list_t init) : list(init, false, false) {}
+};
+
+///
+/// A JSON-like builder for creating arrays.
+///
+class array : public list {
+    using initializer_list_t = std::initializer_list<list>;
+
+   public:
+    ///
+    /// Creates an empty array.
+    ///
+    array() : list({}, false, true){};
+
+    ///
+    /// Creates a BSON array.
+    ///
+    /// @param init
+    ///     the initializer list used to construct the BSON array
+    ///
+    /// @see bsoncxx::builder::list
+    /// @see bsoncxx::builder::document
+    ///
+    array(initializer_list_t init) : list(init, false, true) {}
+};
```

Comment:
As inner classes, these cannot publicly inherit the methods of `builder::list` (i.e., `operator view()` and `view()`). There was no way to get a `bson_value::view` from either class without explicitly converting these classes to a `builder::list`. I realized this when I tried to print documents for the code snippets in the comments of examples. The code snippet below didn't work. 
```
builder::document doc = {"hello", "world"}
std::cout << bsoncxx::to_json(doc.view().get_document()) << std::endl; // No 'view' method for 'builder::document'
```
I updated the test in `bson_builder.cpp` as well.

Reply:
That makes sense. Alternatively, the nested classes could have separate copies of `operator view()` and `view()`, but this seems ok as well.

Reply:
I think I could've. The only issue with that approach is that the nested classes didn't hold on to the underlying `builder::list` or `bson_value::value`. This seemed simpler and represented the as-is relationship a bit better since nested classes don't have very many special privileges in c++.